src/main/scala/ir/transforms/CalleePreservedParam.scala

```diff
@@ -0,0 +1,62 @@
+package ir
+package transforms
+
+import cilvisitor.*
+
+object CalleePreservedParam {
+
+  /**
+   * Asusming single-return and parameter form 
+   */
+
+  val preserved = (19 to 29).map(i => s"R$i").toSet
+
+  def isPreservedParam(v: LocalVar) = {
+    // abi assumption
+    v.varName match {
+      case "R29_in" | "R30_in" | "R31_in" | "R29" | "R30" | "R31" | "R29_out" | "R30_out" | "R31_out" => true
+      case o if preserved.contains(o.stripSuffix("_in").stripSuffix("_out")) => true
+      case _ => false
+    }
+  }
+
+  def transform(p: Program) = {
+    val v = MakePreserved()
+    visit_prog(v, p)
+    assert(invariant.correctCalls(p))
+  }
+
+  class MakePreserved extends CILVisitor {
+
+    override def vproc(p: Procedure) = {
+
+      for (param <- p.formalOutParam.filter(isPreservedParam)) {
+
+        val input = p.formalInParam.find(_.name.startsWith(param.varName.stripSuffix("_out"))).get
```

Comment:
i can't read this. what is it doing?

also, i'm suspicious of using startsWith. for instance, R10 starts with R1, but this is obviously not what you're looking for here (i know that, in this case, all the register numbers are 2 digits, but it still makes me uneasy).